library/blivet.py

```diff
@@ -665,73 +665,101 @@ def _get_size(self):
+        return size
+
+    def _create_vdo(self):
+        parent_device = self._blivet_pool._device
+        size = self._get_size()
+
+        pool_size = parent_device.free_space if size > parent_device.free_space else size
+
+        if self._volume['vdo_pool_size']:
+            try:
+                pool_size = self._trim_size(Size(self._volume['vdo_pool_size']))
+            except BlivetAnsibleError:
+                # Literally the same error just more specific description
+                raise BlivetAnsibleError("specified 'vdo_pool_size' for volume '%s' "
+                                         "exceeds available space in pool '%s' (%s)" % (pool_size,
+                                                                                        parent_device.name,
+                                                                                        parent_device.free_space))
+
+        try:
+            vdopool = self._blivet.new_lv(name='vdopool', vdo_pool=True,
+                                          parents=[parent_device], compression=self._volume['compression'],
+                                          deduplication=self._volume['deduplication'],
+                                          size=pool_size)
+        except Exception as e:
+            raise BlivetAnsibleError("failed to set up VDO pool '%s': %s" % (self._volume['name'], str(e)))
+
+        self._blivet.create_device(vdopool)
+
+        return dict(vdo_lv=True, parents=[vdopool])
+
+    def _create_lvmraid(self):
+        parent_device = self._blivet_pool._device
+
+        if self._volume['raid_disks']:
+            pvs = list()
+
+            for path in self._volume['raid_disks']:
+                for pv in parent_device.pvs:
+                    disk_path = pv.path if pv.is_disk else pv.disk.path
+                    # We need to ensure we match '/dev/sda' with 'sda' but not with 'da'
+                    if ('/' + disk_path).endswith('/' + path):
+                        pvs.append(pv)
+        else:
+            pvs = parent_device.pvs
+
+        return dict(seg_type=self._volume['raid_level'], pvs=pvs)
+
     def _create(self):
         if self._device:
             return
 
-        parent = self._blivet_pool._device
-        if parent is None:
+        parent_device = self._blivet_pool._device
+        if parent_device is None:
             raise BlivetAnsibleError("failed to find pool '%s' for volume '%s'" % (self._blivet_pool['name'], self._volume['name']))
 
-        size = self._get_size()
+        use_vdo = self._volume['deduplication'] or self._volume['compression']
+        use_lvmraid = self._volume['raid_level']
 
-        fmt = self._get_format()
-        trim_percent = (1.0 - float(parent.free_space / size)) * 100
-        log.debug("size: %s ; %s", size, trim_percent)
+        # VDO size is technically unlimited, so no size checks needed when it is used
+        size = self._get_size() if use_vdo else self._trim_size(self._get_size())
```

Comment:
Please use "full" `if-else` here, we don't need to save space and using function call as an function argument is very hard to read and debug, it's not saving worth the two or three extra lines here.

Reply:
Fixed.